### PR TITLE
[custom ops] disable kernel temporarily 

### LINKED ISSUE #128621: Be able to dynamically disable a kernel for a custom op
Useful for when custom ops have a default implementation vs a backend implementation. Sometimes we want to disable the backend implementation for testing/benchmarking purposes.

### PR DESCRIPTION
**Pull request by yushangdi**
Fixes #128621

Sometimes we want to disable the backend implementation for testing/benchmarking purposes.

For example:

```python
@custom_op("mylib::f", mutates_args=())
def f(x: Tensor) -> Tensor:
    return torch.zeros(1)

print(f(torch.randn(1))) # tensor([0.])

@f.register_kernel("cpu")
def _(x):
    return torch.ones(1)

print(f(torch.randn(1))). # tensor([1.])

with f.set_kernel_enabled("cpu", enabled = False):
    print(f(0)) # tensor([0.])
```

### CODE REVIEW

File: torch/_library/custom_ops.py [hunk 1]
```diff
@@ -187,6 +191,54 @@ def _qualname(self) -> str:
     def __repr__(self) -> str:
         return f"<CustomOpDef({self._qualname})>"
 
+    @contextmanager
+    def disable_kernel(self, device_type: str, disable: bool = True):
```
Comment (williamwen42) [inline torch/_library/custom_ops.py:195]: I think it's a little confusing for this to be named `disable_kernel` and also have a `disable` option. Either there should be 2 implementations (e.g. enable, disable), or use a different name (e.g. set_kernel_disabled?). I'm taking inspiration from `torch.no_grad, torch.enable_grad, torch.set_grad_enabled`.

File: torch/_library/custom_ops.py [hunk 2]
```diff
@@ -187,6 +191,54 @@ def _qualname(self) -> str:
...
+            If a kernel is first disabled and then registered, it is disabled until enabled again.
+
+        Args:
+            device_type (str): The device type to disable/enable the kernel for.
+            disable (bool): Whether to disable or enable the kernel.
+        """
+        action = "disable" if disable else "enable"
+        originally_disabled = device_type in self._disabled_kernel
+        if device_type not in self._backend_fns:
+            warnings.warn(
```
Comment (williamwen42) [inline torch/_library/custom_ops.py:211]: Log instead of warning, since warning will only issue once?

File: torch/_library/custom_ops.py (hunk 1 above)
Comment (zou3519) [reply torch/_library/custom_ops.py:195]: Maybe set_kernel_enabled(enabled=True/False) ?

File: torch/_library/custom_ops.py (hunk 2 above)
Comment (yushangdi) [reply torch/_library/custom_ops.py]: changed to use log now.

**Review by williamwen42 — APPROVED** [review]

File: torch/_library/custom_ops.py [hunk 3]
```diff
@@ -187,6 +192,55 @@ def _qualname(self) -> str:
     def __repr__(self) -> str:
         return f"<CustomOpDef({self._qualname})>"
 
+    @contextmanager
+    def set_kernel_enabled(self, device_type: str, enabled: bool = True):
```
Comment (zou3519) [inline torch/_library/custom_ops.py:196]: I don't think we have docs for CustomOpDef on the website. Could you figure out how to add them sometime? (doesn't have to be in this PR).

In the past the APIs we have on CustomOpDef are the same as the ones in torch.library.*, but now we have a CustomOpDef specific API.

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**